ledger/eval/eval.go

```diff
@@ -972,17 +978,39 @@ func (eval *BlockEvaluator) TransactionGroup(txgroup []transactions.SignedTxnWit
 
 	// Evaluate each transaction in the group
 	txibs = make([]transactions.SignedTxnInBlock, 0, len(txgroup))
+	cowForTxn := cow
 	for gi, txad := range txgroup {
 		var txib transactions.SignedTxnInBlock
 
+		if eval.GranularEval {
+			if gi == 0 {
+				cowForTxn = cow.child(1)
+				defer cowForTxn.recycle()
+			} else {
+				// Reuse the same cow allocation for all transactions in this group
+				cowForTxn.reset()
+				cow.reuseChild(cowForTxn, 1)
+			}
+		}
+
```

Comment:
I dislike the increasing complexity around recycling cows.  Unfortunately, I don't have a great suggestion yet, but we've already had bugs in our recycling, and wish we could prevent it from getting more fiddly.

Why can't we recycle the cow after calling AfterTxn? I thought I saw you copying out the state delta.

Reply:
We can recycle after AfterTxn. The code here is an optimization to reuse the same cow for each txn in the group instead of going through the recycling pool.

We can instead achieve the same thing by doing `cowForTxn := cow.child(1)` at the beginning of the loop and `cowForTxn.recycle()` at the end for each iteration (though maybe we'd missing some recycling opportunities if we don't use defer).